src/main/scala/ru/otus/sc/App.scala

```diff
@@ -1,22 +1,92 @@
 package ru.otus.sc
 
-import ru.otus.sc.greet.dao.impl.GreetingDaoImpl
-import ru.otus.sc.greet.model.{GreetRequest, GreetResponse}
-import ru.otus.sc.greet.service.GreetingService
-import ru.otus.sc.greet.service.impl.GreetingServiceImpl
+import ru.otus.sc.greet.dao.ExchangeRate
+import ru.otus.sc.greet.dao.impl.{ExchangeRatesDaoImpl, GreetingDaoImpl}
+import ru.otus.sc.greet.model._
+import ru.otus.sc.greet.service.impl.{
+  EchoServiceImpl,
+  ExchangeRatesServiceImpl,
+  GreetingServiceImpl,
+  InvocationCounterServiceImpl,
+  LocalStoreServiceImpl
+}
+import ru.otus.sc.greet.service.{
+  EchoService,
+  ExchangeRatesService,
+  GreetingService,
+  InvocationCounterService,
+  LocalStoreService
+}
 
 trait App {
   def greet(request: GreetRequest): GreetResponse
+  def getCurrencyExchangeRate(request: ExchangeRatesRequest): ExchangeRatesResponse
+  def echo[T](request: EchoRequest[T]): EchoResponse[T]
 }
 
 object App {
-  private class AppImpl(greeting: GreetingService) extends App {
-    def greet(request: GreetRequest): GreetResponse = greeting.greet(request)
+  private class AppImpl(
+      greeting: GreetingService,
+      exchangeRates: ExchangeRatesService,
+      echoService: EchoService,
+      counterService: InvocationCounterService,
+      localStoreService: LocalStoreService
+  ) extends App {
+    def greet(request: GreetRequest): GreetResponse = {
+      counterService.incrementAndGet(greeting.getServiceName)
+      greeting.greet(request)
+    }
+
+    def getCurrencyExchangeRate(request: ExchangeRatesRequest): ExchangeRatesResponse = {
+      counterService.incrementAndGet(exchangeRates.getServiceName)
+
+      // Сначала пытаемся взять значение из локального кэша
+      localStoreService.get(request.secid) match {
+        case Some(value: ExchangeRate) =>
+          App.log(s"Return rate from cache: $value")
+          ExchangeRatesResponse(Some(value))
+        case _ =>
+          val response: ExchangeRatesResponse = exchangeRates.getExchangeRate(request)
+          // Если удалось получить значение сохраняем его в кэш
+          response.rate match {
+            case Some(value) =>
+              localStoreService.put(value.secid, value) match {
+                case Some(exception) => App.log(exception.getMessage)
+                case None =>
```

Comment:
```suggestion
                case None => ()
```

Если что-то возвращает `Unit`, то лучше быть максимально явным об этом, но ещё лучше стараться избегать сайд-эффектов там где они не нужны.